onnxruntime/core/providers/xnnpack/nn/unpool.cc

```diff
@@ -0,0 +1,266 @@
+  // exception of the batch size. Can be removed once we've run more models using xnnpack MaxUnpool.
+  auto inferred_output_shape = utils::GetTensorShapeFromTensorShapeProto(*Node().OutputDefs()[0]->Shape());
+  ORT_ENFORCE(inferred_output_shape[1] == output_dims_[1] &&
+                  inferred_output_shape[2] == output_dims_[2] &&
+                  inferred_output_shape[3] == output_dims_[3],
+              "Shape mismatch between inferred value and calculated value.");
+  uint32_t input_padding_top = gsl::narrow<uint32_t>(pool_attrs_.pads[0]);
+  uint32_t input_padding_left = gsl::narrow<uint32_t>(pool_attrs_.pads[1]);
+  uint32_t input_padding_bottom = gsl::narrow<uint32_t>(pool_attrs_.pads[2]);
+  uint32_t input_padding_right = gsl::narrow<uint32_t>(pool_attrs_.pads[3]);
+
+  uint32_t pooling_height = gsl::narrow<uint32_t>(pool_attrs_.kernel_shape[0]);
+  uint32_t pooling_width = gsl::narrow<uint32_t>(pool_attrs_.kernel_shape[1]);
+
+  xnn_status status = xnn_status_invalid_state;
+  struct xnn_operator* p = nullptr;
+  op_type_ = OpComputeType::op_compute_type_fp32;
+  status = xnn_create_unpooling2d_nhwc_x32(input_padding_top, input_padding_right,
+                                           input_padding_bottom, input_padding_left,
+                                           pooling_height, pooling_width,
+                                           C, C, C,  // channels, input_pixel_stride, output_pixel_stride
+                                           0, &p);
+  ORT_ENFORCE(status == xnn_status_success, "xnn_create_max_unpooling2d_nhwc_",
+              OpTypeToString(op_type_), "failed. Status:", status);
+
+  op0_.reset(p);
+}
+
+Status MaxUnpool::Compute(OpKernelContext* context) const {
+  const auto& X = *context->Input<Tensor>(0);
+  const auto& indice = *context->Input<Tensor>(1);
+  const auto& X_shape = X.Shape();
+
+  int64_t N = X_shape[0];
+  int64_t H = X_shape[1];
+  int64_t W = X_shape[2];
+
+  // set the N dim to the correct value
+  TensorShapeVector output_dims{output_dims_};
+  output_dims[0] = N;
+
+  if (is_indice_produced_by_xnnpack_ == 0) {
+    if (num_inputs_ == 3) {
+      const auto& output_shape = *context->Input<Tensor>(2);
+      auto output_shape_span = output_shape.DataAsSpan<int64_t>();
+      output_dims[1] = output_shape_span[2];
+      output_dims[2] = output_shape_span[3];
+      output_dims[3] = output_shape_span[1];
+    }
+    Tensor* Y = context->Output(0, output_dims);
+    const int64_t HW = H * W;
+    const int64_t Channel = X_shape[3];
+    const int64_t CHW = Channel * HW;
+    const int64_t oHW = output_dims[1] * output_dims[2];
+    const int64_t oCHW = Channel * oHW;
+
+    const auto* X_data = reinterpret_cast<const int32_t*>(X.DataRaw());
+    const auto* I_data = indice.Data<int64_t>();
+    auto out = gsl::make_span(reinterpret_cast<int32_t*>(Y->MutableDataRaw()), Y->Shape().Size());
+    std::fill_n(out.data(), out.size(), 0);
```

Comment:
nit: assuming you're going to set every element in the output, the fill_n isn't adding value.

Reply:
its signature is `(int* __first, unsigned long __n, const int& __value)`. So it should work as expected?  I put `out.size()` as the second parament but not `out.end()`